GeosCore/tagged_o3_mod.F90

```diff
@@ -365,233 +277,241 @@ SUBROUTINE CHEM_TAGGED_O3( Input_Opt, State_Chm, State_Diag, State_Grid, &
+    ! Check if the HEMCO State object is a valid pointer
+    ! (this only has to be done on the first call)
+    !========================================================================
+    IF ( first ) THEN
 
        ! Make sure the HEMCO State object has been first allocated
-       IF ( .NOT. ASSOCIATED(HcoState) ) THEN
-          ErrMsg = 'The HcoState object is not allocated!'
-          CALL GC_Error( ErrMsg, RC, ThisLoc )
+       IF ( .NOT. ASSOCIATED( HcoState ) ) THEN
+          errMsg = 'The HcoState object is not allocated!'
+          CALL GC_Error( errMsg, RC, thisLoc )
           RETURN
        ENDIF
 
-       ! Get emissions timestep (seconds)
-       TS_EMIS = HcoState%TS_EMIS
-
        ! Reset first-time variable
        ! so that this won't be executed again
-       FIRST    = .FALSE.
-
+       first  = .FALSE.
     ENDIF
 
-    ! Evalulate O3 production from HEMCO
-    ALLOCATE( P24H(State_Grid%NX, State_Grid%NY, State_Grid%NZ), STAT=RC )
-    CALL GC_CheckVar( 'tagged_o3_mod.F: P24H', 0, RC )
-    IF ( RC /= GC_SUCCESS ) RETURN
-    P24H = 0.0_fp
-    CALL HCO_GC_EvalFld( Input_Opt, State_Grid, 'O3_PROD', P24H, RC )
+    !========================================================================
+    ! Get production and loss frequencies from HEMCO. These are read
+    ! from the "ProdLoss" History collection (archived from a fullchem
+    ! simulation) and have units [molec/cm3/s].
+    !========================================================================
+
+    ! P(O3) from HEMCO [molec/cm3/s]
+    CALL HCO_GC_EvalFld( Input_Opt, State_Grid, 'O3_PROD', PO3_hco, RC )
     IF ( RC /= GC_SUCCESS ) THEN
-       ErrMsg = 'Cannot get pointer to O3_PROD!'
-       CALL GC_Error( ErrMsg, RC, ThisLoc )
+       errMsg = 'Cannot get O3_PROD [molec/cm3/s]!'
+       CALL GC_Error( errMsg, RC, thisLoc )
        RETURN
     ENDIF
 
-    ! Set pointer to O3 production
-    P24Hptr => P24H
-
-    ! Evaluate O3 loss from HEMCO
-    ALLOCATE( L24H(State_Grid%NX, State_Grid%NY, State_Grid%NZ), STAT=RC )
-    CALL GC_CheckVar( 'tagged_o3_mod.F: L24H', 0, RC )
-    IF ( RC /= GC_SUCCESS ) RETURN
-    L24H = 0.0_fp
-    CALL HCO_GC_EvalFld( Input_Opt, State_Grid, 'O3_LOSS', L24H, RC )
+    ! L(O3) from HEMCO [molec/cm3/s]
+    CALL HCO_GC_EvalFld( Input_Opt, State_Grid, 'O3_LOSS', LO3_hco, RC )
     IF ( RC /= GC_SUCCESS ) THEN
-       ErrMsg = 'Cannot get pointer to O3_LOSS!'
-       CALL GC_Error( ErrMsg, RC, ThisLoc )
+       errMsg = 'Cannot get O3_LOSS [molec/cm3/s]!'
+       CALL GC_Error( errMsg, RC, thisLoc )
        RETURN
     ENDIF
 
-    ! DT is the ratio of the chemistry and emission time step.
-    ! Use this value to convert from kg/m3 or 1/m3 per emission
-    ! time step to kg/m3 or 1/m3 per chemistry time step.
-    ! (ckeller, 9/17/2014).
-    DT = DTCHEM / TS_EMIS
-
-    !=================================================================
+    !========================================================================
```

Comment:
The comments here (and unit conversions below) state that prod and loss of O3 from HEMCO are in units molec/cm3/s. However, when I do ncdump on the files indicated in HEMCO_Config.rc:

```
* O3_PROD        $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.ProdLoss.$YYYY$MM$DD_0000z.nc4         Prod_Ox           2010-2019/1-12/1/0 C xyz 1        * - 1 1
* O3_LOSS        $ROOT/GCClassic_Output/13.0.0/$YYYY/GEOSChem.ProdLoss.$YYYY$MM$DD_0000z.nc4         Loss_Ox           2010-2019/1-12/1/0 C xyz 1        * - 1 1
```

it seems the units are in fact in kg/s:

```
        float Loss_Ox(time, lev, lat, lon) ;
                Loss_Ox:long_name = "Chemical loss of Ox" ;
                Loss_Ox:units = "kg s-1" ;
                Loss_Ox:averaging_method = "time-averaged" ;

        float Prod_Ox(time, lev, lat, lon) ;
                Prod_Ox:long_name = "Chemical production of Ox" ;
                Prod_Ox:units = "kg s-1" ;
                Prod_Ox:averaging_method = "time-averaged" ;
```

Also, does it matter that technically these are prod/loss rates of Ox (as defined in fullchem.kpp) and not O3 that we are reading in via HEMCO?

Reply:
I think it doesn't strictly matter that we use Prod_Ox and Loss_Ox, as these are the families that are saved out by KPP.  Ox is mostly O3 anyway.  We would have to rebuild the fullchem mechanism to add P(O3) and L(O3).

Also, I wonder if something changed between 13.0.0 and 13.4.0.  In my testing, I used 1-day output from a fullchem run and the Prod_Ox and Loss_Ox are definitely in molec/cm3/s:

```console
        float Loss_Ox(time, lev, lat, lon) ;
                Loss_Ox:long_name = "Chemical loss of Ox" ;
                Loss_Ox:units = "molec cm-3 s-1" ;
                Loss_Ox:averaging_method = "time-averaged" ;
                Loss_Ox:_Storage = "chunked" ;
                Loss_Ox:_ChunkSizes = 1, 1, 46, 72 ;
                Loss_Ox:_DeflateLevel = 1 ;
                Loss_Ox:_Shuffle = "true" ;
                Loss_Ox:_Endianness = "little" ;
                Loss_Ox:_NoFill = "true" ;
         float Prod_Ox(time, lev, lat, lon) ;
                Prod_Ox:long_name = "Chemical production of Ox" ;
                Prod_Ox:units = "molec cm-3 s-1" ;
                Prod_Ox:averaging_method = "time-averaged" ;
                Prod_Ox:_Storage = "chunked" ;
                Prod_Ox:_ChunkSizes = 1, 1, 46, 72 ;
                Prod_Ox:_DeflateLevel = 1 ;
                Prod_Ox:_Shuffle = "true" ;
                Prod_Ox:_Endianness = "little" ;
                Prod_Ox:_NoFill = "true" ;
```
I'm not sure if `Hco_EvalFld` can let you check the units of the data.  Maybe we should only convert units if the data is in molec/cm3/s.

Reply:
I just found this issue #965, which fixed a problem in the ProdLoss units.  The data in the 13.0.0 GCClassic output is in molec/cm3/s but the units are kg/s.  So I think that's OK.